Token propagation E2E tests.

rubenvp8510 opened the pull request:
Signed-off-by: Ruben Vargas <ruben.vp8510@gmail.com>

~~Also, in my tests I realized that  this is not gonna work if the query service has` --es.use-aliase` flag enabled.~~

heisenberg8510 commented:
In the end, I haven't need to move anything on the image, it works if I add to the "test" openshift user the cluster-admin-role, this works because the way the https://github.com/fabric8io/openshift-elasticsearch-plugin works.

If the `subjectAccessReview` is satisfied, it assigns a role to the user, in this case in the` origin-logging-elasticsearch5` image there is a jaeger configuration for that, this allows some users to have access to the indices. (and that is why the cluster-admin can do requests ). This will allow us to test the whole thing, E2E test :) 

I will update my tests here (and update the branch) .. after that this is ready for review.

rubenvp8510 commented:
In the end, I haven't needed to move anything on the image, it works if I add to the `test` openshift user the c`luster-admin-role,` this does the job because the way the `openshift-elasticsearch-plugin` works[1].

If the `subjectAccessReview` is satisfied, it assigns a role to the user, in this case in the` origin-logging-elasticsearch5` image there is a jaeger[2] configuration for that. This allows some users to have access to the indices. (and that is why the cluster-admin can do requests :D ). This will allow us to test the whole thing without touch the image.. I think this works at least for testing... 

1) https://github.com/fabric8io/openshift-elasticsearch-plugin 
2) https://github.com/openshift/origin-aggregated-logging/blob/master/elasticsearch/sgconfig/sg_config.yml#L21

jpkrohling commented:
Is this ready for review? If so, there are a couple of things that needs to be cleaned up first, like commented out code and the user change in the `Makefile`.

rubenvp8510 commented:
@jpkrohling  Ahh! there is small details that I need to fix, (one of them is the things you pointed in your comment) ;)

Thanks

rubenvp8510 commented:
@jpkrohling now is ready

jpkrohling commented on pkg/util/util.go:
```diff
@@ -198,3 +200,17 @@ func InitObjectMeta(obj metav1.Object) {
+	_, err := rand.Read(randString)
+	if err != nil {
+		// If we cannot generate random, return fixed.
+		return "ncNDoqLGrayxXzxTn5ANbOXZp3qXd0LA"
```
Return an error instead, and let the caller decide. At the very least, there should be a WARN in the logs about this. It was really bad that I didn't realize that "SECRET" should have been a real random string...

jpkrohling commented on pkg/inject/oauth_proxy.go:
```diff
@@ -44,6 +45,15 @@ func getOAuthProxyContainer(jaeger *v1.Jaeger) corev1.Container {
+		args = append(args,
+			"--pass-access-token=true",
+			"--pass-user-bearer-token=true",
+			"--scope=user:info user:check-access user:list-projects",
```
Spaces as separators? That's odd!

jpkrohling commented on .ci/start-openshift.sh:
```diff
@@ -9,7 +9,10 @@ export KUBERNETES_VERSION=v1.14.0
 sudo docker cp $(docker create docker.io/openshift/origin:$OPENSHIFT_VERSION):/bin/oc /usr/local/bin/oc
-oc cluster up --version=$OPENSHIFT_VERSION
+
+oc cluster up --version=$OPENSHIFT_VERSION --public-hostname=${HOST_IP}.nip.io
```
cc @kevinearls

kevinearls commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,278 @@
+
+	client, err := oAuthAuthorization(queryHost, "user-test-token", "any")
+
+	require.NoError(t, err)
```
The test is failing for me here when I use a real OCP 4.2 OpenShift cluster.  Have you been testing against minishift?  I think you probably need to use the actual route here.

rubenvp8510 replied:
I haven't tested against OCP 4.2/minishift, I'll give it a try and fix what is failing. Thanks!

objectiser commented on pkg/apis/jaegertracing/v1/jaeger_types.go:
```diff
@@ -177,6 +177,9 @@ type JaegerQuerySpec struct {
 	Options Options `json:"options,omitempty"`
 
+	// +optional
+	TokenPropagation bool `json:"tokenPropagation,omitempty"`
```
Not sure we should add this to the `JaegerQuerySpec` yet - as it is already available via the `Options`. We might want to in the future, but unless there is a specific clear benefit of promoting this option, when we haven't for others, then it is better for now to leave it.

rubenvp8510 replied:
Ok right, this is only for testing and won't be available to general use. I'll remove it and use options,

pavolloffay replied:
+1 on removal for now

objectiser commented on pkg/strategy/production.go:
```diff
@@ -105,8 +105,8 @@ func newProductionStrategy(jaeger *v1.Jaeger, es *storage.ElasticsearchDeploymen
 
-			es.InjectStorageConfiguration(&queryDep.Spec.Template.Spec)
-			es.InjectStorageConfiguration(&cDep.Spec.Template.Spec)
+			es.InjectStorageConfiguration(&queryDep.Spec.Template.Spec, !query.TokenPropagation())
```
Why is `TokenPropagation` configuration being used to determine is `tls` is enabled? Surely they are separate concerns - i.e. `tls` may be enabled regardless of whether the token is propagated or not?

rubenvp8510 replied:
Because if TLS is enabled (tls authentication), the token will be ignored by ES.

as far as I know, this is not disabling TLS communication, but authentication

objectiser replied:
Unless I am misunderstanding something, the `tls` is protecting the transport/communications between the collector and ES, regardless of the originating user - whereas the token is specific to the user making the request (used for authorization) - so they are two different things?

rubenvp8510 replied:
Not exactly, the `tls` flag  enable/disable ssl client authentication. if you disable `tls` but specify a CA, the communication will be protected .

You can see the behaviour here:

https://github.com/jaegertracing/jaeger/blob/master/pkg/es/config/config.go#L294
https://github.com/jaegertracing/jaeger/blob/master/pkg/es/config/config.go#L340

Said that, I think the flag name is a little bit confusing.

pavolloffay replied:
@rubenvp8510 could you please document this in `InjectStorageConfiguration`?

pavolloffay commented on pkg/deployment/query.go:
```diff
@@ -183,3 +185,8 @@ func (q *Query) labels() map[string]string {
 	return fmt.Sprintf("%s-query", q.jaeger.Name)
 }
+
+//TokenPropagation returns true is token propagation is enabled on query service.
```
s/is/if

pavolloffay commented on pkg/inject/oauth_proxy.go:
```diff
@@ -33,16 +33,36 @@ func OAuthProxy(jaeger *v1.Jaeger, dep *appsv1.Deployment) *appsv1.Deployment {
+}
+
+func getOAuthProxyContainer(jaeger *v1.Jaeger) corev1.Container {
+
```
could you please remove empty lines?

pavolloffay commented on pkg/inject/oauth_proxy.go:
```diff
@@ -33,16 +33,36 @@ func OAuthProxy(jaeger *v1.Jaeger, dep *appsv1.Deployment) *appsv1.Deployment {
+	secret, err := util.GenerateProxySecret()
+	if err != nil {
+		jaeger.Logger().Warnf("Error generating secret: %s, fallback to fixed secret", secret)
+		secret = "ncNDoqLGrayxXzxTn5ANbOXZp3qXd0LA"
```
Could you please explain this? What this secret stands for?

pavolloffay replied:
nit export it to a constant with a comment what does it stand for

rubenvp8510 replied:
Secret should be something random, but in case of failure I just fallback to something fixed, now the operator uses the word "SECRET"

pavolloffay commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,284 @@
+		return true, nil
+	})
+	require.NoError(t, err, "Token propagation test failed")
+
```
nit: empty line

pavolloffay commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,284 @@
+}
+
+func NewHTTPSClient() *http.Client {
+
```
nit: remove empty lines in this func

pavolloffay commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,284 @@
+	return exampleJaeger
+}
+
+func NewHTTPSClient() *http.Client {
```
Do you need to export this?

pavolloffay commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,284 @@
+}
+
+func oAuthAuthorization(host, user, pass string) (*http.Client, error) {
+
```
nit: remove empty lines in this func

rubenvp8510 commented:
> The test does not look sufficient - just to assert on the status code from query service. Could you please assert on the exact content returned from the service? e.g. create a span and try to

rubenvp8510 commented:
At this point I'm wondering if this test is necessary or is enough with the test implemented on jaeger [1]. This is due the fact that for now the operator won't have official support for this.

https://github.com/jaegertracing/jaeger/pull/1705

objectiser commented:
@rubenvp8510 I think the test is required - it performs an e2e test in a real environment with actual ES. The referenced PR just uses a mock ES.

rubenvp8510 commented:
> @rubenvp8510 I think the test is required - it performs an e2e test in a real environment with actual ES. The referenced PR just uses a mock ES.

My only concern is the fact that we are testing a feature here than is not part of the operator, so the failure of this test does not mean anythingin the operator context.

Anyway it is just an opinion, I'll address all comments and update this PR. :)

rubenvp8510 commented:
I removed the `tokenPropagation` from jaeger spec, but I need to add the `Options` attribute to the Ingress so I can add more arguments to oauth proxy, Also added the possibility to disable query TLS client authentication against ES.

I did this test more complete, now besides of check the http status of the response I insert an span and expect it on the results.

@jpkrohling @pavolloffay @objectiser Please review it :)

objectiser commented on pkg/storage/elasticsearch.go:
```diff
@@ -38,6 +38,53 @@ type ElasticsearchDeployment struct {
+	if strings.EqualFold(util.FindItem("--es-archive.enabled", container.Args), "--es-archive.enabled=true") {
+		container.Args = append(container.Args,
+			"--es-archive.server-urls="+elasticsearchURL,
+			"--es-archive.tls=true",
```
Shouldn't this use the same override approach as for the `es.tls` parameter?

rubenvp8510 replied:
For this test is not necessary, but yeah for consistency I can use the same approach.

objectiser commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,356 @@
+
+func (suite *TokenPropagationTestSuite) SetupSuite() {
+	t = suite.T()
+	if !isOpenShift(t) {
```
Is there a reason this is openshift (es-operator) specific? Shouldn't it also work with any external ES cluster?

rubenvp8510 replied:
The test is openshift specific, but it can work with an external ES cluster. but I assumed that the whole test environment is openshift, or at least the oauth part used with oauth proxy sidecar.

jpkrohling replied:
The OAuth part is OpenShift-specific, but if we can make this test work without the OAuth Proxy, then we can test it in the Kubernetes test suite.

objectiser commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,356 @@
+	namespace, _ = ctx.GetNamespace()
+	require.NotNil(t, namespace, "GetNamespace failed")
+	addToFrameworkSchemeForSmokeTests(t)
+
```
nit: remove line.

objectiser commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,356 @@
+	handleSuiteTearDown()
+}
+
+func (suite *TokenPropagationTestSuite) TestTokenPropagationTest() {
```
nit: Remove `Test` at the end of the name?

rubenvp8510 replied:
Hahaha, redundant, I'll remove.

objectiser commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,356 @@
+
+	client, err := oAuthAuthorization(host, username, password)
+
+	err = wait.Poll(retryInterval, timeout, func() (done bool, err error) {
```
This test is showing a successful response with token propagation. Can we also have a test that is expected to fail - i.e. propagates an invalid token?

rubenvp8510 replied:
Yes sure, in that way we can do the test with an invalid token or without the token at all. and we should get a forbidden error.

jpkrohling replied:
Those tests seem to be missing.

objectiser commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -119,6 +118,20 @@ func (suite *TokenPropagationTestSuite) TestTokenPropagationTest() {
 		Finish()
 	closer.Close()
 
+	client := &http.Client{
```
Could this check be moved into a separate test - makes test failure reporting better.

jpkrohling commented on .ci/run-e2e-tests.sh:
```diff
@@ -39,6 +39,16 @@ elif [ "${TEST_GROUP}" = "examples2" ]
+    oc create user user-test-token
+    oc adm policy add-cluster-role-to-user cluster-admin user-test-token
+    # for ocp 4.2
+    htpasswd -c -B -b users.htpasswd user-test-token any
```
It's a better idea to just ship the actual file instead of generating the file at the CI.

jpkrohling commented on .ci/run-e2e-tests.sh:
```diff
@@ -39,6 +39,16 @@ elif [ "${TEST_GROUP}" = "examples2" ]
+then
+    echo "Running token propagation tests"
+    oc create user user-test-token
+    oc adm policy add-cluster-role-to-user cluster-admin user-test-token
```
The bad thing about using "root-like" permissions in a test is that you might miss the cases where the user might not have the appropriate permissions. Not sure it's relevant for your test here, but in general, we should aim to have the minimum set of permissions possible.

jpkrohling commented on pkg/inject/oauth_proxy.go:
```diff
@@ -33,17 +35,26 @@ func OAuthProxy(jaeger *v1.Jaeger, dep *appsv1.Deployment) *appsv1.Deployment {
+	}
 	args := []string{
-		"--cookie-secret=SECRET",
+		fmt.Sprintf("--cookie-secret=%s", secret),
```
This generates secrets containing chars like "+" and "/". How are they translated into the final YAML? Do they cause any troubles? I don't think it does, but it's good to have positive confirmation (and perhaps a test). Example of random strings I got from this code locally:

```
pZ8F0imysFsTwJ9kF8U92g==
ILmvfqWW/+bLEokY06NHig==
Zmc+jX2VWXkxsB98AIYwDg==
xb/fYVT6UHxYpNJXhjR7RQ==
```

jpkrohling commented on pkg/storage/elasticsearch.go:
```diff
@@ -38,6 +38,53 @@ type ElasticsearchDeployment struct {
 }
 
+func (ed *ElasticsearchDeployment) injectArguments(container *corev1.Container) {
+	container.Args = append(container.Args,
```
Probably a matter of taste, but do you need to split this line here? The next line seems small enough to fit here.

rubenvp8510 replied:
no I don't need to split the line not sure why I did it, or may be the make format did it for me? Anyway I will change it.

jpkrohling commented on pkg/storage/elasticsearch.go:
```diff
@@ -38,6 +38,53 @@ type ElasticsearchDeployment struct {
+		container.Args = append(container.Args, "--es.tls=true")
+	}
+
+	container.Args = append(container.Args,
```
Shouldn't these options be added only if `es.tls=true`?

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,397 @@
+
+}
+
+func (suite *TokenPropagationTestSuite) TearDownSuite() {
```
Do you need this func if it's empty? The commented out code should be removed anyway.

rubenvp8510 replied:
Ah this should not be empty, I commented those lines on my local but shouldn't be commented.

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,397 @@
+	}
+
+	err := wait.Poll(retryInterval, timeout, func() (done bool, err error) {
+		req, err := http.NewRequest(http.MethodGet, suite.queryServiceEndPoint, nil)
```
Check that error is nil, and return early in case of errors (`return false, err`). Otherwise, it will fail in the next line.

jpkrohling replied:
Looks like this wasn't done yet.

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,397 @@
+
+	/* Try to reach query endpoint */
+	err = wait.Poll(retryInterval, timeout, func() (done bool, err error) {
+		req, err := http.NewRequest(http.MethodGet, suite.queryServiceEndPoint, nil)
```
Same comment as before: make the test more resilient, otherwise it might become annoying to all future PRs ;)

jpkrohling replied:
You probably want to check the `err` objects in this method and log their causes, without hard failing.

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,397 @@
+		Jar: cookieJar,
+	}
+	/* Start oauth */
+	resp, err := client.Get("https://" + host + "/oauth/start")
```
Kinda of the same as previous comments, but you probably just want to retry in case of some well-known networking problems, like timeouts.

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,397 @@
+	var req *http.Request
+	/* Submit form */
+	if hasForm(responseBytes) {
+		req = getLoginFormRequest(responseBytes, resp.Request.URL, user, pass)
```
For OpenShift, do you really need to parse the form and simulate a real user there? Can't you just send the user:pass with basic auth? I think the Maistra components to send the requests via basic auth.

rubenvp8510 replied:
Well, I can get a token an pass it to query service, but I don't know how to do it and test the whole thing using oauth proxy sidecar, Can we use basic authentication to and set the oauth proxy cookie? May be I'm missing something. Or we don't need to test using the sidecar?

jpkrohling replied:
Maistra QE can confirm, but I understand that you can send basic auth and the proxy will do all the token negotiation based on it. Jaeger will see a token, just like if it was a browser session.

jpkrohling commented on pkg/inject/oauth_proxy.go:
```diff
@@ -33,17 +36,26 @@ func OAuthProxy(jaeger *v1.Jaeger, dep *appsv1.Deployment) *appsv1.Deployment {
+func proxyInitArguments(jaeger *v1.Jaeger) []string {
+	secret, err := util.GenerateProxySecret()
+	if err != nil {
+		jaeger.Logger().Warnf("Error generating secret: %s, fallback to fixed secret", secret)
```
`WithError(err)`, so that the cause is logged.

jpkrohling commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,337 @@
+			APIVersion: "jaegertracing.io/v1",
+		},
+	}))
+	assert.NoError(t, framework.AddToFrameworkScheme(apis.AddToScheme, &esv1.ElasticsearchList{
```
Perhaps you want to use `require` here instead of `assert`?

jpkrohling commented:
@kevinearls could you give this a test in OpenShift before this gets merged?

kevinearls commented:
@jpkrohling Sure, I'll try it later today or first thing tomorrow

kevinearls commented:
@jpkrohling @rubenvp8510  I've tried this on an OCP 4.2 cluster, but unfortunately have not been able to get it to work so far.  The TestTokenPropagationNoToken test passes, but TestTokenPropagationNoToken fails at line 134 with a 403.  

I will note a couple of issues I came across in the code, but I don't have any ideas so far on how to get this to work.

rubenvp8510 commented:
@kevinearls I'll check, TestTokenPropagationNoToken should throws a 403 (Forbidden) , is the expected thing, so I will check why test fails.

kevinearls commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,337 @@
+		require.Equal(t, http.StatusOK, resp.StatusCode)
+		if resp.StatusCode != http.StatusOK {
+			return false, errors.New("Query service returns http code: " + string(resp.StatusCode))
+		}
```
The if is redundant here as if the require.Equal fails the test will terminate.  You should pick one or the other.  In general I prefer the require as it makes for more concise code and produces better error messages.  However, if you prefer the if you need to change the string(resp.StatusCode) to strconv.Itoa

rubenvp8510 replied:
I preferred required too, but I had to do in this way, so if it fails the first time due the pod not fully ready or other timing issues, the wait.Poll will retry again.

If I use required, it will fail mediately. Making this test less reliable. IMHO.

I removed the required line.

kevinearls commented on test/e2e/elasticsearch_token_propagation_test.go:
```diff
@@ -0,0 +1,337 @@
+}
+
+func bindOperatorWithAuthDelegator() {
+	roleBinding := rbac.ClusterRoleBinding{
```
This clusterrolebinding and the token-test-user-bind clusterrolebinding created below need to be deleted at the end of the test (in TearDownSuite()) otherwise the test will fail after the first run as these will already exist on the cluster.

rubenvp8510 replied:
Thanks for the observation, I've added cleaning routine for remove all created roles/role-bindings.

kevinearls commented:
@rubenvp8510 Sorry, my comment about the test failure was wrong.  TestTokenPropagationValidToken is the test that fails, TestTokenPropagationNoToken works correctly.

Sorry about the careless cut and paste on my part.  :anguished:

rubenvp8510 commented:
I ran the tests a couple of times, and sometimes fails and others not, I'll do this more reliable

jpkrohling commented:
Thanks for being persistent, @rubenvp8510!
